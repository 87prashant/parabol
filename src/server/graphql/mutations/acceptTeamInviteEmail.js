import {GraphQLID, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import parseInviteToken from 'server/graphql/mutations/helpers/inviteTeamMembers/parseInviteToken';
import validateInviteTokenKey from 'server/graphql/mutations/helpers/inviteTeamMembers/validateInviteTokenKey';
import NotifyAddedToTeam from 'server/graphql/types/NotifyAddedToTeam';
import acceptTeamInvite from 'server/safeMutations/acceptTeamInvite';

export default {
  type: new GraphQLNonNull(NotifyAddedToTeam),
  description: `Add a user to a Team given an invitationToken.
    If the invitationToken is valid, returns the auth token with the new team added to tms.
    Side effect: deletes all other outstanding invitations for user.`,
  args: {
    inviteToken: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The invitation token (first 6 bytes are the id, next 8 are the pre-hash)'
    }
  },
  async resolve(source, {inviteToken}, {authToken, dataLoader, socketId}) {
    const r = getRethink();
    const now = new Date();
    const operationId = dataLoader.share();

    // VALIDATION
    const {id: inviteId, key: tokenKey} = parseInviteToken(inviteToken);

    // see if the invitation exists
    const invitation = await r.table('Invitation').get(inviteId).update({
      tokenExpiration: new Date(0),
      updatedAt: now
    }, {returnChanges: true})('changes')(0)('old_val').default(null);

    if (!invitation) {
      throw new Error('notFound');
    }

    const {tokenExpiration, hashedToken, teamId, email} = invitation;
    // see if the invitation has expired
    if (tokenExpiration < now) {
      throw new Error('expiredInvitation');
    }

    // see if the invitation hash is valid
    const isCorrectToken = await validateInviteTokenKey(tokenKey, hashedToken);
    if (!isCorrectToken) {
      throw new Error('invalidToken');
    }
    const oldtms = authToken.tms || [];
    // Check if TeamMember already exists (i.e. user invited themselves):
    const teamMemberExists = oldtms.includes(teamId);
    if (teamMemberExists) {
      throw new Error('alreadyJoined');
    }

    // RESOLUTION
    return acceptTeamInvite(teamId, authToken, email, {operationId, mutatorId: socketId});
  }
};

