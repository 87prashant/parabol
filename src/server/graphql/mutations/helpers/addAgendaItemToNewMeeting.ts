import {AGENDA_ITEMS} from 'universal/utils/constants'
import getRethink from 'server/database/rethinkDriver'
import {DataLoaderWorker} from 'server/graphql/graphql'
import Meeting from 'server/database/types/Meeting'
import AgendaItemsPhase from 'server/database/types/AgendaItemsPhase'
import AgendaItemsStage from 'server/database/types/AgendaItemsStage'

/*
 * NewMeetings have a predefined set of stages, we need to add the new agenda item manually
 */
const addAgendaItemToNewMeeting = async (
  agendaItemId,
  teamId: string,
  dataLoader: DataLoaderWorker
) => {
  const now = new Date()
  const r = getRethink()
  const team = await dataLoader.get('teams').load(teamId)
  const {meetingId} = team
  if (!meetingId) return undefined
  // make sure it's a new meeting
  const newMeeting = (await r
    .table('NewMeeting')
    .get(meetingId)
    .default(null)) as Meeting
  if (!newMeeting) return undefined
  const {phases} = newMeeting
  const agendaItemPhase = phases.find((phase) => phase.phaseType === AGENDA_ITEMS) as
    | AgendaItemsPhase
    | undefined
  if (!agendaItemPhase) return undefined

  const {stages} = agendaItemPhase
  const newStage = new AgendaItemsStage(agendaItemId)
  newStage.isNavigable = true
  newStage.isNavigableByFacilitator = true
  stages.push(newStage)
  await r
    .table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    })
  return meetingId
}

export default addAgendaItemToNewMeeting
