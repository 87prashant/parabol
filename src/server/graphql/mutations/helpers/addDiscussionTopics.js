import getRethink from 'server/database/rethinkDriver';
import shortid from 'shortid';
import {DISCUSS, TEAM} from 'universal/utils/constants';
import publish from 'server/utils/publish';
import StartNewMeetingPayload from 'server/graphql/types/StartNewMeetingPayload';


export const makeDiscussionStage = (reflectionGroupId, meetingId) => ({
  id: shortid.generate(),
  meetingId,
  isComplete: false,
  phaseType: DISCUSS,
  reflectionGroupId
});

const mapGroupsToStages = (reflectionGroups) => {
  const importantReflectionGroups = reflectionGroups.filter((group) => group.voterIds.length > 0);
  // handle edge case that no one votes
  if (importantReflectionGroups.length === 0) return reflectionGroups;
  importantReflectionGroups.sort((a, b) => a.voterIds.length < b.voterIds.length ? 1 : -1);
  return importantReflectionGroups;
};

const addDiscussionTopics = async (meeting, dataLoader, subOptions) => {
  const r = getRethink();
  const now = new Date();
  const {phases, teamId} = meeting;
  const discussPhase = phases.find((phase) => phase.phaseType === DISCUSS);
  if (!discussPhase) return false;
  const {id: meetingId} = meeting;
  const reflectionGroups = await dataLoader.get('retroReflectionGroupsByMeetingId').load(meetingId);
  const importantReflectionGroups = mapGroupsToStages(reflectionGroups);
  const discussStages = importantReflectionGroups.map((reflectionGroup) => makeDiscussionStage(reflectionGroup.id, meetingId));

  if (!discussStages[0] || !discussPhase.stages[0]) return false;
  // mutative, replaces the placeholder with an actual stage
  discussStages[0].id = discussPhase.stages[0].id;
  discussPhase.stages = discussStages;

  await r.table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    });
  const data = {teamId, meetingId};
  publish(TEAM, teamId, StartNewMeetingPayload, data, {operationId: subOptions.operationId});
  return true;
};

export default addDiscussionTopics;
