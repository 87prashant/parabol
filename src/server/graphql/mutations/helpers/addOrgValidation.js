import legitify from 'universal/validation/legitify'
import {fullName, orgName, requiredEmail, teamName} from 'universal/validation/templates'

export default function addOrgValidation () {
  return legitify({
    invitees: [
      {
        email: requiredEmail,
        fullName
      }
    ],
    newTeam: {
      name: teamName
    },
    orgName
  })
}
