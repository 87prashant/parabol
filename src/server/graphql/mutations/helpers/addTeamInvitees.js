import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';

const addTeamInvitees = async (invitees, teamId, viewerId) => {
  if (!invitees || invitees.length === 0) return {teamInviteNotifications: [], invitationIds: []};
  const {teamInviteNotifications, newInvitations} = await inviteTeamMembers(invitees, teamId, viewerId);
  const invitationIds = newInvitations.map(({id}) => id);
  return {teamInviteNotifications, invitationIds};
};

export default addTeamInvitees;
