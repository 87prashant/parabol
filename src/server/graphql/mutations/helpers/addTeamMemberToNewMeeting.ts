import createMeetingMembers from 'server/graphql/mutations/helpers/createMeetingMembers'
import {CHECKIN, UPDATES} from 'universal/utils/constants'
import getRethink from 'server/database/rethinkDriver'
import CheckInStage from 'server/database/types/CheckInStage'
import UpdatesStage from 'server/database/types/UpdatesStage'
import CheckInPhase from 'server/database/types/CheckInPhase'
import UpdatesPhase from 'server/database/types/UpdatesPhase'

/*
 * NewMeetings have a predefined set of stages, we need to add the new team member manually
 */

const setInPhase = (phase: CheckInPhase | UpdatesPhase, newStage: CheckInStage | UpdatesStage) => {
  const firstStage = phase.stages[0]
  newStage.isNavigable = firstStage.isNavigable
  newStage.isNavigableByFacilitator = firstStage.isNavigableByFacilitator
  phase.stages.push(newStage)
}

const addTeamMemberToNewMeeting = async (teamMember, teamId, dataLoader) => {
  const now = new Date()
  const r = getRethink()
  const team = await dataLoader.get('teams').load(teamId)
  const {meetingId} = team
  if (!meetingId) return false
  // make sure it's a new meeting
  const newMeeting = await r
    .table('NewMeeting')
    .get(meetingId)
    .default(null)
  if (!newMeeting) return false
  const {phases} = newMeeting
  const checkInPhase = phases.find((phase) => phase.phaseType === CHECKIN)
  const updatesPhase = phases.find((phase) => phase.phaseType === UPDATES)
  if (checkInPhase) {
    setInPhase(checkInPhase, new CheckInStage(teamMember.id))
  }
  if (updatesPhase) {
    setInPhase(updatesPhase, new UpdatesStage(teamMember.id))
  }
  const [meetingMember] = await createMeetingMembers(newMeeting, [teamMember], dataLoader)
  await r({
    meeting: r
      .table('NewMeeting')
      .get(meetingId)
      .update({
        phases,
        updatedAt: now
      }),
    member: r.table('MeetingMember').insert(meetingMember)
  })
  return true
}

export default addTeamMemberToNewMeeting
