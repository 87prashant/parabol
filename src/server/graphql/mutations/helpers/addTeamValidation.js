import legitify from 'universal/validation/legitify';
import {requiredId, fullName, requiredEmail, teamName} from 'universal/validation/templates';

export default function addTeamValidation() {
  return legitify({
    invitees: [{
      email: requiredEmail,
      fullName
    }],
    newTeam: {
      name: teamName,
      orgId: requiredId
    }
  });
}
