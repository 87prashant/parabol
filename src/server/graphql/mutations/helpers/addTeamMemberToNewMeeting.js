import {CHECKIN} from 'universal/utils/constants';
import getRethink from 'server/database/rethinkDriver';
import {makeCheckInStage} from 'server/graphql/mutations/helpers/makeCheckinStages';

/*
 * NewMeetings have a predefined set of stages, we need to add the new team member manually
 */
const addTeamMemberToNewMeeting = async (teamMember, teamId, dataLoader) => {
  const now = new Date();
  const r = getRethink();
  const team = await dataLoader.get('teams').load(teamId);
  const {meetingId} = team;
  if (!meetingId) return false;
  // make sure it's a new meeting
  const newMeeting = await r.table('NewMeeting').get(meetingId).default(null);
  if (!newMeeting) return false;
  const {phases} = newMeeting;
  const checkInPhase = phases.find((phase) => phase.phaseType === CHECKIN);
  if (!checkInPhase) return false;

  const {stages} = checkInPhase;
  const newStage = makeCheckInStage(teamMember, meetingId, false);
  stages.push(newStage);
  await r.table('NewMeeting')
    .get(meetingId)
    .update({
      phases,
      updatedAt: now
    });
  return true;
};

export default addTeamMemberToNewMeeting;
