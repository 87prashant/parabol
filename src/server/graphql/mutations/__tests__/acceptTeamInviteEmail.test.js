import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockPubSub from 'server/__mocks__/MockPubSub';
import makeDataLoader from 'server/__tests__/setup/makeDataLoader';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import * as adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import hashInviteTokenKey from 'server/graphql/models/Invitation/inviteTeamMembers/hashInviteTokenKey';
import makeInviteToken from 'server/graphql/models/Invitation/inviteTeamMembers/makeInviteToken';
import parseInviteToken from 'server/graphql/models/Invitation/inviteTeamMembers/parseInviteToken';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {ADD_USER} from 'server/utils/serverConstants';
import * as tmsSignToken from 'server/utils/tmsSignToken';
import {TEAM_INVITE} from 'universal/utils/constants';
import acceptTeamInviteEmail from 'server/graphql/mutations/acceptTeamInviteEmail';

MockDate.set(__now);
console.error = jest.fn();

describe('acceptTeamInviteEmail', () => {
  beforeEach(() => {
    auth0ManagementClient.users.updateAppMetadata.mockReset();
  });

  test.only('adds an invitee who was not previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const mockPubSub = new MockPubSub();
    const invitee = newInvitee('acceptTeamInviteEmail1');
    const inviteToken = makeInviteToken();
    const {id} = parseInviteToken(inviteToken);
    const hashedToken = await hashInviteTokenKey(inviteToken);
    await mockDB.init()
      .newUser({name: 'inviteeGuy', tms: [], userOrgs: [], email: invitee.email})
      .newNotification(undefined, {type: TEAM_INVITE, email: invitee.email})
      .newInvitation({id, hashedToken, email: invitee.email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.context.user;
    const authToken = mockAuthToken(inviteeUser);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [teamId];
    // TEST
    await acceptTeamInviteEmail.resolve(undefined, {inviteToken}, {authToken, dataLoader});

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation')
        .getAll(inviteeUser.email, {index: 'email'})
        .filter({teamId})
        .orderBy('tokenExpiration')
        .coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledWith(inviteeUser.id, orgId, ADD_USER);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
  test('adds an invitee who was previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const mockPubSub = new MockPubSub();
    const inviteToken = makeInviteToken();
    const {id} = parseInviteToken(inviteToken);
    const hashedToken = await hashInviteTokenKey(inviteToken);
    await mockDB.init()
      .newTeam({orgId: mockDB.context.organization.id})
      .newNotification(undefined, {type: TEAM_INVITE, email: mockDB.db.user[1].email})
      .newInvitation({id, hashedToken, email: mockDB.db.user[1].email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.db.user[1];
    const authToken = mockAuthToken(inviteeUser);
    const dataLoader = makeDataLoader(authToken);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [mockDB.db.team[0].id, teamId];

    // TEST
    await acceptTeamInviteEmail.resolve(undefined, {inviteToken}, {authToken, dataLoader});

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation')
        .getAll(inviteeUser.email, {index: 'email'})
        .filter({teamId})
        .orderBy('tokenExpiration')
        .coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledTimes(0);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
});
