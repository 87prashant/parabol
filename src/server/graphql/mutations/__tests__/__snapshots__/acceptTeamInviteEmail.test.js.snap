// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acceptTeamInviteEmail adds an invitee who was not previously in the org 1`] = `
Object {
  "invitation": Array [
    Object {
      "acceptedAt": 2017-01-01T08:00:00.000Z,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "<<constant>>",
      "id": "<<constant>>",
      "inviteCount": 1,
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 1970-01-01T00:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 4,
    "email": "invitee@mockTeam+acceptTeamInviteEmail1.co",
    "id": 3,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "teamId": 1,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "invitee@mockTeam+acceptTeamInviteEmail1.co",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "inviteeGuy",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "tms": Array [
      1,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 2,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`acceptTeamInviteEmail adds an invitee who was previously in the org 1`] = `
Object {
  "invitation": Array [
    Object {
      "acceptedAt": 2017-01-01T08:00:00.000Z,
      "createdAt": 2017-01-01T07:00:00.000Z,
      "email": "<<constant>>",
      "hashedToken": "$2a$10$HdSFfY82SEVkDLrU8We6Bu561wdfYAIKRvH4g3lqAfuUVfqV8k1RW",
      "id": "<<constant>>",
      "inviteCount": 1,
      "inviteToken": "qQrg64bVC6xe9z",
      "invitedBy": "<<constant>>",
      "teamId": "<<constant>>",
      "tokenExpiration": 1970-01-01T00:00:00.000Z,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 6,
    "email": "vision@avengers.org",
    "id": 5,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "teamId": 2,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "vision@avengers.org",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Jonas",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "tms": Array [
      1,
      2,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 3,
        "role": null,
      },
      Object {
        "id": 4,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`acceptTeamInviteEmail adds an invitee who was previously in the org 2`] = `
Object {
  "invitationRemoved": Array [
    Object {
      "channelId": 2,
      "message": Object {
        "invitationRemoved": Object {
          "invitation": Object {
            "acceptedAt": 2017-01-01T08:00:00.000Z,
            "createdAt": 2017-01-01T07:00:00.000Z,
            "email": 8,
            "hashedToken": 9,
            "id": 7,
            "inviteCount": 1,
            "inviteToken": 10,
            "invitedBy": 11,
            "teamId": 2,
            "tokenExpiration": 1970-01-01T00:00:00.000Z,
            "updatedAt": 2017-01-01T08:00:00.000Z,
          },
        },
        "operationId": 0,
      },
    },
  ],
  "newAuthToken": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "newAuthToken": 12,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 2,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "preferredName": "Vision",
              "teamId": 2,
              "teamName": "Team America",
              "type": "JOIN_TEAM",
            },
          ],
        },
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsCleared": Object {
          "deletedIds": Array [
            13,
          ],
        },
      },
    },
  ],
  "teamAdded": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "mutatorId": undefined,
        "operationId": 0,
        "teamAdded": Object {
          "notification": Object {
            "startAt": 2017-01-01T08:00:00.000Z,
            "teamId": 2,
            "type": "ADD_TO_TEAM",
          },
          "teamId": 2,
          "teamMemberId": 5,
        },
      },
    },
  ],
  "teamMemberAdded": Array [
    Object {
      "channelId": 2,
      "message": Object {
        "mutatorId": undefined,
        "operationId": 0,
        "teamMemberAdded": Object {
          "teamMember": Object {
            "checkInOrder": 1,
            "email": 8,
            "id": 5,
            "isCheckedIn": null,
            "isFacilitator": true,
            "isLead": false,
            "isNotRemoved": true,
            "picture": "https://placeimg.com/100/100/animals",
            "preferredName": "Vision",
            "teamId": 2,
            "userId": 0,
          },
        },
      },
    },
  ],
}
`;
